bitcoin/src/bip32.rs

```diff
@@ -730,33 +1076,73 @@ impl Xpub {
     /// the internal public key representation.
     pub fn to_x_only_pub(self) -> XOnlyPublicKey { XOnlyPublicKey::from(self.public_key) }
 
+    /// Derive an extended public key from a path that contains NormalChildNumbers.
+    pub fn derive_pub<
+        'a,
+        C: secp256k1::Verification,
+        T: IntoIterator<Item = &'a NormalChildNumber>,
+    >(
+        &self,
+        secp: &Secp256k1<C>,
+        path: T,
+    ) -> Xpub {
+        let mut pk: Xpub = *self;
+        for normal_cnum in path.into_iter() {
+            pk = pk.internal_derive_pub(secp, normal_cnum.to_number()).unwrap();
+        }
+        pk
+    }
+
     /// Attempts to derive an extended public key from a path.
     ///
-    /// The `path` argument can be any type implementing `AsRef<ChildNumber>`, such as `DerivationPath`, for instance.
-    pub fn derive_pub<C: secp256k1::Verification, P: AsRef<[ChildNumber]>>(
+    /// The `path` argument can be any type implementing `IntoIterator<Item = &'a ChildNumber>`, such as `DerivationPath`, for instance.
+    pub fn try_derive_pub<
+        'a,
+        C: secp256k1::Verification,
+        P: IntoIterator<Item = &'a ChildNumber>,
+    >(
         &self,
         secp: &Secp256k1<C>,
-        path: &P,
+        path: P,
     ) -> Result<Xpub, Error> {
         let mut pk: Xpub = *self;
-        for cnum in path.as_ref() {
-            pk = pk.ckd_pub(secp, *cnum)?
+        for normal_cnum in path.into_iter() {
+            pk = pk.internal_derive_pub(secp, *normal_cnum)?
         }
         Ok(pk)
     }
 
+    /// Public->Public child key derivation
+    fn internal_derive_pub<C: secp256k1::Verification, N: TryInto<ChildNumber>>(
+        &self,
+        secp: &Secp256k1<C>,
+        number_from: N,
+    ) -> Result<Xpub, Error> {
```

Comment:
```suggestion
    ) -> Result<Xpub, N::Error> {
```

This should fix it. I don't see anything else obviously wrong. If Rustc says ambiguous type try `<N as TryInto>::Error`

Reply:
`.map_err(|_| Error::CannotDeriveFromHardenedKey)?;` fixed

Reply:
No, that's not correct, you need to do it as I wrote to be able to prove infallibility. But now I see the second reason it was that you call other fallible methods with `?` operator. The first one should not return result because the number is guaranteed to be normal - you need to refactor `ckd_pub_tweak` method too. The second is cryptographically impossible, so you can unwrap it.

Reply:
That's a nice idea, but since were using the `TryInto<ChildNumber>` we cannot ensure nor check if the number is a normal one... in the marjority of the time were calling `internal_derive_pub` with `NormalChildNumber.to_number`.
Im changing it to `TryInto<NormalChildNumber>` to avoid unecessary conversion dand ensure that the number is normal

Reply:
Crap, did I suggest using `TryInto<ChildNumber>`? Because I definitely meant to say `TryInto<NormalChildNumber>`.

Reply:
Yes, [Here](https://github.com/rust-bitcoin/rust-bitcoin/pull/3036#discussion_r1688405595).